# DATA 1 Practical 2 - Questions

Simos Gerasimou


## Classic Cars & Co

Classic Cars & Co is a UK company that has a large collection of classic cars from the 1980s. 

DataVision (the company you are working as a Data Scientist) has been contracted to analyse the data available for the cars and provide insights by analysing the different characteristics of the cars (e.g., speed, price). 

This Jupyter Notebook will be presented to the Classic Cars & Co main stakeholders who have limited knowledge about data science. So, your findings should be complemented by a suitable justification explaining what you observe and, when applicable, what does this observation mean and, possibly, why it occurs. The analysis along with the explanation will help them to understand whether they need to invest more to expand their collection.

### **Important Information**

(1) To answer these exercises, you **must first read Chapter 2: Introduction to NumPy from the Python Data Science Handbook** (https://jakevdp.github.io/PythonDataScienceHandbook/02.00-introduction-to-numpy.html)


(2) For each question (task) a description is provided accompanied (most of the time) by two cells: one for writing the Python code and another for providing the justification. Feel free to add more cells if you feel they are needed, but keep the cells corresponding to the same question close by.

**Hint 1**: If you find difficulties in solving a task, look at Chapter 2 from the Python Data Science Handbook.

**Hint 2**: Solving each task using NumPy should require less than 10 lines of code

#### **T1) Explore the dataset and for each column write its name, data type (categorical/numerical - nominal,ordinal,discrete,continuous) and its meaning (i.e., what does it capture?)**

* You may want to open the CSV file using a text editor (e.g., Notepad) or a spreadsheet editor (e.g., Excel)

**Write your answer here (the first is given)**
* Make: Categorical (Nominal) - The model of the car
* ....

### 1) Reading dataset

The classic cars dataset is available on VLE (look for classicCars.csv in the Practicals section)

In [15]:
#Using NumPy to read the dataset
import numpy as np
import pandas as pd
import math
#Define the path to the dataset
data_path = "ClassicCars.csv"
#Define the type of each dataset column. 
#This is needed because NumPy arrays cannot directly read files with different data types
#Hence, we are using Structured arrays. 
#But, we will soon move to Pandas which makes data manipulation easier
types = ['U20', 'U10', 'U5', 'U20', 'U3', 'f4', 'f4', 'f4', 'f4', 'U10', 'i4', 'i4', 'i4', 'i4', 'i4']
#Read the dataset
data = np.genfromtxt(data_path, dtype=types, delimiter=',', names=True)
df = pd.read_csv(data_path); df

make fueltype numofdoors    bodystyle drivewheels  wheelbase  \
0    alfa-romero      gas        two  convertible         rwd       88.6   
1    alfa-romero      gas        two  convertible         rwd       88.6   
2    alfa-romero      gas        two    hatchback         rwd       94.5   
3           audi      gas       four        sedan         fwd       99.8   
4           audi      gas       four        sedan         4wd       99.4   
..           ...      ...        ...          ...         ...        ...   
200        volvo      gas       four        sedan         rwd      109.1   
201        volvo      gas       four        sedan         rwd      109.1   
202        volvo      gas       four        sedan         rwd      109.1   
203        volvo   diesel       four        sedan         rwd      109.1   
204        volvo      gas       four        sedan         rwd      109.1   

     length  width  height numofcylinders  enginesize  horsepower  citympg  \
0     168.8   64.1    48.8           four         130         111       21   
1     168.8   64.1    48.8           four         130         111       21   
2     171.2   65.5    52.4            six         152         154       19   
3     176.6   66.2    54.3           four         109         102       24   
4     176.6   66.4    54.3           five         136         115       18   
..      ...    ...     ...            ...         ...         ...      ...   
200   188.8   68.9    55.5           four         141         114       23   
201   188.8   68.8    55.5           four         141         160       19   
202   188.8   68.9    55.5            six         173         134       18   
203   188.8   68.9    55.5            six         145         106       26   
204   188.8   68.9    55.5           four         141         114       19   

     highwaympg  price  
0            27  13495  
1            27  16500  
2            26  16500  
3            30  13950  
4            22  17450  
..          ...    ...  
200          28  16845  
201          25  19045  
202          23  21485  
203          27  22470  
204          25  22625  

[205 rows x 15 columns]

**Structured Arrays**
* Read more about structured arrays:
  * https://jakevdp.github.io/PythonDataScienceHandbook/02.09-structured-data-numpy.html
  * https://numpy.org/doc/stable/user/basics.rec.html

### Analysing the dataset


#### **Extracting the column names**

In [13]:
data.dtype.names

('make',
 'fueltype',
 'numofdoors',
 'bodystyle',
 'drivewheels',
 'wheelbase',
 'length',
 'width',
 'height',
 'numofcylinders',
 'enginesize',
 'horsepower',
 'citympg',
 'highwaympg',
 'price')

#### **Extracting the shape of the array**

In [14]:
print("The shape of the array is: ", data.shape)
print(data.)

SyntaxError: invalid syntax (<ipython-input-14-ed0cb195be74>, line 2)

#### **T2) What do you see?**
* How many entries does the array have?
* What does each entry include? 
* Hint: Print the elements of an entry


**Write your answer here**

205 entries
make,fueltype,numofdoors,bodystyle,drivewheels,wheelbase,length,width,height,numofcylinders,enginesize,horsepower,citympg,highwaympg,price

#### **Extracting the entries of a column given its name**

* By specifying the name of a column, you can get all the entries within the array for this column (reminder: you are using Structured Arrays)


In [ ]:
#Print the entries within the 'make' column
print(data['make'])

#### **T3) Extract the bodystyles within the dataset**


In [11]:
print(data['bodystyle'])


['convertible' 'convertible' 'hatchback' 'sedan' 'sedan' 'sedan' 'sedan'
 'wagon' 'sedan' 'hatchback' 'sedan' 'sedan' 'sedan' 'sedan' 'sedan'
 'sedan' 'sedan' 'sedan' 'hatchback' 'hatchback' 'sedan' 'hatchback'
 'hatchback' 'hatchback' 'hatchback' 'sedan' 'sedan' 'sedan' 'wagon'
 'hatchback' 'hatchback' 'hatchback' 'hatchback' 'hatchback' 'hatchback'
 'sedan' 'wagon' 'hatchback' 'hatchback' 'sedan' 'sedan' 'sedan' 'sedan'
 'sedan' 'sedan' 'sedan' 'hatchback' 'sedan' 'sedan' 'sedan' 'hatchback'
 'hatchback' 'hatchback' 'sedan' 'sedan' 'hatchback' 'hatchback'
 'hatchback' 'hatchback' 'hatchback' 'sedan' 'hatchback' 'sedan' 'sedan'
 'hatchback' 'sedan' 'sedan' 'sedan' 'wagon' 'hardtop' 'sedan' 'sedan'
 'convertible' 'sedan' 'hardtop' 'hatchback' 'hatchback' 'hatchback'
 'hatchback' 'hatchback' 'hatchback' 'hatchback' 'hatchback' 'hatchback'
 'hatchback' 'sedan' 'sedan' 'sedan' 'sedan' 'sedan' 'sedan' 'sedan'
 'sedan' 'wagon' 'sedan' 'hatchback' 'sedan' 'wagon' 'hardtop' 'hatchback'
 'seda

### How do the car prices look like?


#### **T4) Calculate the range of car prices for the entire dataset**


In [21]:
#Write your answer here
print(np.ptp(data['price']))


40282
40282


#### **T5) Calculate the min, max, mean and median prices of the cars**


In [20]:
#Write your answer here
maximum = np.max(data['price'])
minimum = np.min(data['price'])
mean = np.sum(data['price']) // len(data['price'])
median = np.sort(data['price'])[len(data['price'])//2]
print(maximum, minimum, mean, median)


45400 5118 13300 10345


#### **T6) Considering the values calculated above, what insights can you extract? Where do you think the majority of car prices will be clustered?**


**Write your answer here**

The majority of carprices are low, with the median being just above 10,000 and the mean at just above 13,000 this shows the general price of the car is quite low

#### **T7) Write code to calculate the standard deviation for the car prices. Then use the corresponding NumPy function to confirm the correctness of your calculation**



In [27]:
#Write your answer here
answer = np.std(data['price'])
print(answer)

7969.54140103854


#### **T8) Find the details of cars with the smallest and largest car volumes**
* Hint: see how to calculate the volume of a car https://info.japanesecartrade.com/content-item/297-what-is-m3-cubic-meter-size-of-a-vehicle


In [50]:
#Write your answer here
#make,fueltype,numofdoors,bodystyle,drivewheels,wheelbase,length,width,height,numofcylinders,enginesize,horsepower,citympg,highwaympg,price
tmp = []
max_volume = np.max(data['length']/100*data['width']/100*data['height']/100)*10
min_volume = np.min(data['length']/100*data['width']/100*data['height']/100)*10
print(min_volume, max_volume) 

4.526431560516357 8.460077047348022


#### **T9) Find the different types of bodystyles for the cars in the dataset**

* Hint: You may want to check: https://numpy.org/doc/stable/reference/generated/numpy.unique.html

In [51]:
#Write your answer here
print(np.unique(data['bodystyle']))


['convertible' 'hardtop' 'hatchback' 'sedan' 'wagon']


#### **T10) Find the number of different car *brands* (makes)**


In [52]:
#Write your answer here
print(len(np.unique(data['make'])))


22


#### **T11) Find the engine size and the horsepower for the most and least efficient cars when driven in the city and the highway (i.e., the cars with the smallest and largest difference in fuel consumption when driven in the city and the highway)**

In [63]:
#Write your answer here
least_index = np.argmax(data['citympg']-data['highwaympg'])
most_index = np.argmin(data['citympg']-data['highwaympg'])
print(f"""
Most efficient horsepower : {data[most_index]['horsepower']}
Most efficient engine size : {data[most_index]['enginesize']}
Least efficient horsepower : {data[least_index]['horsepower']}
Least efficient engine size : {data[least_index]['enginesize']}
""")



Most efficient horsepower : 288
Most efficient engine size : 203
Least efficient horsepower : 95
Least efficient engine size : 152



#### **T12) Find the make with the largest number of cars and how many they are**

In [16]:
#Write your answer here
unique, pos = np.unique(data['make'], return_inverse=True)
counts = np.bincount(pos)
max_makes = counts.argmax()
print(unique[max_makes], counts[max_makes])


toyota 32


#### **T13) Find how many cars have a wheel base greater than 100**

* Hint: See https://jakevdp.github.io/PythonDataScienceHandbook/02.06-boolean-arrays-and-masks.html

In [17]:
#Write your answer here
import pandas as pd
wheelbases = pd.read_csv('ClassicCars.csv')['wheelbase'].values
np.count_nonzero(wheelbases<88)


2

#### **T14) Find if there are any convertible cars that cost less than £15000**

* Hint: See https://jakevdp.github.io/PythonDataScienceHandbook/02.06-boolean-arrays-and-masks.html

In [9]:
#Write your answer here



#### **T15) Calculate the interquartile range for the price of all cars**

In [10]:
#Write your answer here
q75, q25 = np.percentile(data['price'], [75, 25])
iqr = q75-q25
iqr

8715.0

#### **T16) Calculate the 50th percentile range for the horsepower of all cars. Which value is the 50th percentile equal to?**

In [11]:
#Write your answer here
p50 = np.percentile(data['price'], [50])
p50

array([10345.])

### Ideas for practicing further at home

* Find the engine and horsepower of 4wd cars
* Find whether diesel or gas cars are more efficient in the city/highway
* Any other analysis that you might could generate some useful insight.
